cli/util/upload_bin.go

```diff
@@ -263,131 +267,190 @@ func fetchLatestMissingBlockIndex(ctx context.Context, p *pool.Pool, containerID
+							if errUpload != nil {
+								return errUpload
+							}
+							if debug {
+								fmt.Fprintf(ctx.App.Writer, "Uploaded block %d with object ID: %s\n", blockIndex, resOid.String())
+							}
 							return errUpload
+						}, maxRetries)
+						if errRetr != nil {
+							select {
+							case errCh <- errRetr:
+							default:
+							}
+							return
 						}
-						if debug {
-							fmt.Fprintf(ctx.App.Writer, "Uploaded block %d with object ID: %s\n", blockIndex, resOid.String())
-						}
-						return errUpload
-					}, maxRetries)
-					if errRetr != nil {
-						select {
-						case errCh <- errRetr:
-						default:
-						}
-						return
+						resOid.Encode(buffer[blockIndex%int(indexFileSize)*oidSize:])
 					}
-				}
-			}(i)
-		}
+				}(i)
+			}
 
-		go func() {
-			wg.Wait()
-			close(doneCh)
-		}()
+			go func() {
+				wg.Wait()
+				close(doneCh)
+			}()
 
-		select {
-		case err := <-errCh:
-			return fmt.Errorf("upload error: %w", err)
-		case <-doneCh:
-		}
+			select {
+			case err := <-errCh:
+				return fmt.Errorf("upload error: %w", err)
+			case <-doneCh:
+			}
 
-		fmt.Fprintf(ctx.App.Writer, "Successfully uploaded batch of blocks: from %d to %d\n", batchStart, batchEnd-1)
+			fmt.Fprintf(ctx.App.Writer, "Successfully processed batch of blocks: from %d to %d\n", batchStart, batchEnd-1)
+			batchStart = batchEnd
+		}
+		err := checkBuffer(buffer, uint(indexFileStart/int(indexFileSize)), uint(indexFileEnd-indexFileStart), indexFileSize)
+		if err != nil {
+			return err
+		}
+		if indexFileEnd-indexFileStart == int(indexFileSize) {
+			attrs := []object.Attribute{
+				*object.NewAttribute(indexAttributeKey, strconv.Itoa(indexFileStart/int(indexFileSize))),
+				*object.NewAttribute("IndexSize", strconv.Itoa(int(indexFileSize))),
+			}
+			err = retry(func() error {
+				var errUpload error
+				_, errUpload = uploadObj(ctx.Context, p, signer, acc.PrivateKey().GetScriptHash(), containerID, buffer, attrs, homomorphicHashingDisabled)
+				return errUpload
+			}, maxRetries)
+			if err != nil {
+				return fmt.Errorf("failed to upload index file: %w", err)
+			}
+			fmt.Println("Successfully uploaded index file ", indexFileStart/int(indexFileSize))
+		}
+		clear(buffer)
 	}
 	return nil
 }
 
 // uploadIndexFiles uploads missing index files to the container.
-func uploadIndexFiles(ctx *cli.Context, p poolWrapper, containerID cid.ID, account *wallet.Account, signer user.Signer, oldestMissingBlockIndex uint, blockAttributeKey string, homomorphicHashingDisabled bool, maxParallelSearches, maxRetries int, debug bool) error {
+// It returns the number of uploaded index files and the buffer of  with the uploaded OIDs.
+func uploadIndexFiles(ctx *cli.Context, p poolWrapper, containerID cid.ID, account *wallet.Account, signer user.Signer, oldestMissingBlockIndex, indexFileSize uint, blockAttributeKey, attributeKey string, homomorphicHashingDisabled bool, maxParallelSearches, maxRetries int, debug bool) (uint, []byte, error) {
```

Comment:
Make this function to rely on fact that index files are uploaded in accordance with blocks. Then this function should only search for the latest index file and fill buffer with entries that are already uploaded to the storage. Rename this function accordingly.

Reply:
> files are uploaded in accordance with blocks.

Basically, it is implemented, but also I took into account the fact that we do have the `--index-file-size` and `--index-attribute` flags, which means we can create new index-files based on already uploaded blocks at a random point.

Reply:
So I think the ability to create index files on an already uploaded chain should be saved.

Reply:
Why do we need it?

Reply:
Flexibility and variability.
It would make the `--index-file-size` and `--index-attribute` flags, as well as the `--block-attribute` flag (partually), redundant. Also we do have `IndexFileSize: 128000` configurable in blockfetcher config- means new index files are possible. For example we find out that NeoFS faster with big files and IndexFileSize 1mln is better.

Reply:
Okay, for code and pipeline simplicity I agree. But we should keep in mind that for one block sequence one type of index files. If we want new index files- we need to upload full blocks sequence from the beginning. Will add this to docs.

Reply:
>  If we want new index files- we need to upload full blocks sequence from the beginning.

That's the most significant argument to me. But at the same time maintaining this additional index file generation code becomes a problem, this function has a lot of duplicating code (comparing to `uploadBlocks`), and right now it seems that we don't really need this functionality. We can always get it back (and review/test properly) if we need to resize our index files (or even move it to a separate CLI command), but for now I'd stick to more simple solution.